ea-factories/src/factories/balance.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  ExecuteFactory,
+  Config,
+  DataResponse,
+  Account,
+  SequenceResponseData,
+} from '@chainlink/types'
+import objectPath from 'object-path'
+import { Requester, Validator, AdapterError } from '@chainlink/external-adapter'
+
+const DEFAULT_DATA_PATH = 'addresses'
+const DEFAULT_CONFIRMATIONS = 6
+
+const WARNING_UNSUPPORTED_PARAMS = 'No Operation: this provider does not support'
+const ERROR_MISSING_ADDRESS = 'No Operation: address param is missing'
+
+function groupBy<K, V>(list: Array<V>, keyGetter: (input: V) => K): Map<K, Array<V>> {
+  const map = new Map<K, Array<V>>()
+  list.forEach((item) => {
+    const key = keyGetter(item)
+    const collection = map.get(key)
+    if (!collection) {
+      map.set(key, [item])
+    } else {
+      collection.push(item)
+    }
+  })
+  return map
+}
+
+export function addresses(accounts: Account[]) {
+  return accounts.reduce<string[]>((accumulator, current) => {
+    if (!current.warning) accumulator.push(current.address)
+    return accumulator
+  }, [])
+}
```

Comment:
By using reduce it only iterates over the accounts one time. 

Are you suggesting this because of readability?

Reply:
Yeah, readability! It's one of the most important and hard to achieve things. Naming, readability, composability...
Always think about the next guy, and better yet think about yourself debugging the code a few months down the road, and make it easy for yourself 🥇

Optimization is not even to be considered until the end, or when performance problems arise.

And here we do `O(n)` work in both cases. It is the same if you:

- iterate N elements once, do two operations per element
- iterate N elements twice, do one operation per element

Both are `O(2*n)`, or `O(n)`.

Even if you introduce an extra fixed amount of work because of readability, that is also ok as long as you don't change the time complexity. Going from `O(n)` to `O(6*n)` because of readability is ok while going to `O(n^2)` is not ok, and readability should suffer if that is the tradeoff we need to make in that case.

Reply:
Very true. I'm still in the bad habit of doing things "fancy" sometimes. Blanked on dropping the coefficient :sweat_smile: 

Great advice that should be added to your next book! :book:

Reply:
Haha, don't worry, I only record YouTube videos nowadays. Books are boring and you can't put ads in them 😇